perceval/utils/density_matrix.py

```diff
@@ -0,0 +1,363 @@
+        :param A: any kind of sparse matrix
+        :param val: the array of eigen_values
+        :param vec: the array of eigen_vector
+        """
+        if val.shape[0] != vec.shape[1]:
+            raise ValueError("inconsistent number of eigenvectors and eigenvalues")
+
+        if vec.shape[0] != A.shape[0]:
+            raise ValueError("the size of the matrix is inconsistent with this of the eigenvector")
+
+        def matrix_vector_multiplication(x: np.ndarray):
+            """
+            The matrix vector multiplication function associated to the deflated operator
+            """
+            result = A @ x
+            for k in range(val.shape[0]):
+                result -= val[k] * (conj(vec[:, k].T) @ x) * vec[:, k]
+
+            return result
+
+        return matrix_vector_multiplication
+
+    @staticmethod
+    def _bra_str(bs: BasicState):
+        return "<" + str(bs)[1:][:-1] + "|"
+
+    def to_svd(self, threshold: float = 1e-8, batch_size: int = 1):
+        """
+                gives back an SVDistribution from the density_matrix
+        """
+        if self.size < 50:  # if the matrix is small: array eigh method
+            matrix = self.mat.toarray()
+            w, v = eigh(matrix)
+            dic = {}
+            for k in range(w.shape[0]):
+                sv = array_to_statevector(v[:, k], self.reverse_index)
+                if w[k] > threshold:
+                    dic[sv] = w[k]
+        else:  # if the matrix is large: sparse eigsh method
+            val = np.array([])
+            vec = np.empty(shape=(self._size, 0), dtype=complex)
+            while (val > threshold).all():
+                deflated_operator = LinearOperator((self._size, self._size), matvec=self._deflation(self.mat, val, vec))
+                new_val, new_vec = eigsh(deflated_operator, batch_size)
+                val = np.concatenate((val, new_val))
+                vec = np.concatenate((vec, new_vec), axis=1)
+
+            dic = {}
+            for i in range(val.shape[0]):
+                if val[i] >= threshold:
+                    sv = StateVector()
+                    for j in range(len(vec)):
+                        sv += complex(vec[j][i])*self.reverse_index[j]
+                    sv.normalize()
+                    if sv.m != 0:
+                        dic[sv] = val[i]
+                else:
+                    continue
+        return SVDistribution(dic)
+
+    def __add__(self, other):
+        """
+        add two density Matrices together
+        """
+
+        if not isinstance(other, DensityMatrix):
+            raise TypeError("You can only add a Density Matrix to a Density Matrix")
+
+        if not self._m == other._m:
+            raise ValueError("You can't add Density Matrices acting on different numbers of mode")
+
+        n = max(self._n_max, other.n_max)
+
+        if n == self.n_max:
```

Comment:
Can't you refactor this bit of code ? Since it's almost the same.
Maybe by doing a method that will take the bigger matrix in first argument and the smaller as a second (so you will just have to permute the argument between self and other to make the right matrix)